src/goto-symex/goto_symex_state.cpp

```diff
@@ -664,6 +682,20 @@ static bool requires_renaming(const typet &type, const namespacet &ns)
   return false;
 }
 
+/// Return the type corresponding to a symbol or tag. Empty optional if \p type
+/// is neither a symbol or a tag.
+static optionalt<typet>
+type_of_symbol_or_tag(const typet &type, const namespacet &ns)
```

Comment:
I think you could replace all this with `type = ns.follow(type);`, at the top of `rename_type`, as that's a no-op for non-tags.